src/rust/engine/task_executor/src/lib.rs

```diff
@@ -200,4 +205,112 @@ impl Executor {
       workunit_store::scope_task_workunit_store_handle(workunit_store_handle, future).await
     })
   }
+
+  /// Return a reference to this executor's runtime handle.
+  pub fn handle(&self) -> &Handle {
+    &self.handle
+  }
+}
+
+/// Store "tail" tasks which are async tasks that can execute concurrently with regular
+/// build actions. Tail tasks block completion of a session until all of them have been
+/// completed (subject to a timeout).
+#[derive(Clone)]
+pub struct TailTasks {
+  inner: Arc<Mutex<Option<TailTasksInner>>>,
+}
+
+struct TailTasksInner {
+  id_to_name: HashMap<Id, String>,
+  task_set: JoinSet<()>,
+}
+
+impl TailTasks {
+  pub fn new() -> Self {
+    Self {
+      inner: Arc::new(Mutex::new(Some(TailTasksInner {
+        id_to_name: HashMap::new(),
+        task_set: JoinSet::new(),
+      }))),
+    }
+  }
+
+  /// Spawn a tail task with the given name.
+  pub fn spawn_on<F>(&self, name: &str, task: F, handle: &Handle)
```

Comment:
Nit: Placing a future last in the argument list can make for cleaner callsites (ditto functions):
```rust
tail_tasks.spawn_on(name, handle, async move {
  ..
})
```

Reply:
Will do.

Reply:
Done.